bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+            } else {
+                Err(ApiError::UnexpectedResponse {
+                    from,
+                    state: self.state.name(),
+                    request_id,
+                    msg: "Pkg",
+                })
+            }
+        }
+    }
+
+    // Handle a `ResponseType::Share` from a peer
+    fn on_share(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+        share: Share,
+    ) -> Result<Option<ApiOutput>, ApiError> {
+        match &mut self.state {
+            // We don't send `GetShare` requests in these states
+            State::Uninitialized | State::Learning => {
+                return Err(ApiError::UnexpectedResponse {
+                    from,
+                    state: self.state.name(),
+                    request_id,
+                    msg: "Share",
+                });
+            }
+            State::InitialMember { pkg, distributed_shares } => {
+                validate_share(&from, &share, &pkg.share_digests)?;
+                match self.request_manager.on_share(from, request_id, share) {
+                    Some(TrackableRequest::LoadRackSecret { acks, .. }) => {
+                        let secret = combine_shares(&pkg.share, acks)?;
+                        Ok(Some(ApiOutput::RackSecret { request_id, secret }))
+                    }
+                    Some(TrackableRequest::LearnReceived {
+                        from,
+                        acks,
+                        ..
+                    }) => {
+                        let rack_secret = combine_shares(&pkg.share, acks)?;
+                        // We now have the rack secret and can decrypt extra shares
+                        decrypt_and_send_share_response(
+                            from,
+                            request_id,
+                            pkg,
+                            distributed_shares,
+                            &rack_secret,
+                            &mut self.envelopes,
+                        )
+                    }
+                    // Only LoadRackSecret and LearnReceived track shares so we
+                    // cannot get another variant back.
+                    //
+                    // If we get a `None` back we either haven't received enough
+                    // shares or we have a late response to a prior request. A
+                    // late response is very common, as we terminate the request
+                    // once a threshold is received but we may still receive
+                    // shares because we asked more than a threshold of peers
+                    // for a share. Logging this as an unexpected response would
+                    // be noisy and misleading.
+                    _ => Ok(None),
+                }
+            }
+            State::Learned { pkg } => {
+                validate_share(&from, &share, &pkg.share_digests)?;
+                match self.request_manager.on_share(from, request_id, share) {
```

Comment:
Same thing here; I kinda want this `match` to be hidden inside `request_manager`.

Reply:
Same as above. I'm sure there is likely a 3rd abstraction that would make these pieces fit together a little better, but I'd prefer to not worry about it given urgency.